MLT-Lab-CSE6024_Lab-Assignment-1:
    A Simple ML Model
    
Submitted By: 20MAI0001 - NIHARIKA   MAITRA

Git Repo link: https://github.com/Niharika-20-MAI-01/Winter-2021-MLTechniques-LabCSE6024-Niharika-20MAI01/blob/main/MLT-lab-CSE6024_20MAI0001_A-Simple-ML-Model-LabAssignment-1.ipynb

In [325]:
#Importing necessay Libraries and Packages
import sklearn.datasets
import numpy as np

In [326]:
#Loading our Input Dataset in Jupyter Notebook
cancer_ds = sklearn.datasets.load_breast_cancer()

In [327]:
# Organizing the loaded dataset as X (Feature Vector) and Y (Target Variable)
X = cancer_ds.data
Y = cancer_ds.target

In [328]:
print(X)
print(Y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [329]:
print(X.shape,Y.shape) #569 rows 30 columns present in Feature Vector(X) and 569 rows present in Target Variable(Y)

(569, 30) (569,)


In [330]:
#To get the dataset inside a dataframe using Pandas Library
import pandas as pd
data = pd.DataFrame(cancer_ds.data, columns = cancer_ds.feature_names)

In [331]:
#To Rename the Y variable as Target
data['class'] = cancer_ds.target

In [332]:
#To examine the first few records of our Input dataset
data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

In [333]:
# Describing each row of the Dataset including both Feature Vector(X) and Target(Y) in terms of different Statistical measures/parameters
data.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [334]:
#To find out how many records are malignant and how many are benign in the Input dataset
data['class'].value_counts() #Binary Variable
print(data['class'].value_counts())

1    357
0    212
Name: class, dtype: int64


In [335]:
# To Know the target names or the Labels of Classification Model
print(cancer_ds.target_names)

['malignant' 'benign']


In [336]:
#To decribe the Statistical measure of Mean for different parametes of all the features of the dataset grouped by the Class/Label they belong to i.e, either 0 or 1
data.groupby('class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

# Train Test Split

In [337]:
#Importing train_test_split() function from Sklearn library
from sklearn.model_selection import train_test_split

In [338]:
#Dividing the Input dataset into Feature Vector(X) and Target Variable(Y)
X = data.drop('class', axis = 1)
Y = data['class']

In [339]:
#Line of code to split the dataset in Training and Testing data
X_train, X_test, Y_train, Y_test = train_test_split(X,Y)

In [340]:
# To print the shape of Y(Target Variable) and shape of both the Training and Testing datsets for Y(Target Variable)
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (426,) (143,)


In [341]:
#To print the Mean of Y(Target Variable) and Mean of both the Training and Testing datsets for Y(Target Variable)
print(Y.mean(), Y_train.mean(), Y_test.mean())

0.6274165202108963 0.6338028169014085 0.6083916083916084


In [342]:
#Splitting the Input into Training(90%) and Testing(10%) datasets and using Stratify to push same number of labels into  the training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.1, stratify = Y)
print(X_train.mean(), X_test.mean(), X.mean())

mean radius                 14.132787
mean texture                19.353750
mean perimeter              92.014316
mean area                  655.512500
mean smoothness              0.096430
mean compactness             0.104800
mean concavity               0.089499
mean concave points          0.049027
mean symmetry                0.181332
mean fractal dimension       0.062856
radius error                 0.405836
texture error                1.237711
perimeter error              2.877008
area error                  40.405990
smoothness error             0.007085
compactness error            0.025915
concavity error              0.032414
concave points error         0.011898
symmetry error               0.020688
fractal dimension error      0.003863
worst radius                16.248430
worst texture               25.752852
worst perimeter            107.166367
worst area                 877.756055
worst smoothness             0.132171
worst compactness            0.255238
worst concav

In [343]:
#Splitting the Input into Training(90%) and Testing(10%) datasets and using random_state fixed to particular seed value so that the same record is pushed into the Training and Testing Sets at each execution 
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.1, stratify = Y, random_state = 3)
print(X_train.mean(), X_test.mean(), X.mean())

mean radius                 14.022119
mean texture                19.286211
mean perimeter              91.266348
mean area                  645.180078
mean smoothness              0.096583
mean compactness             0.104042
mean concavity               0.088249
mean concave points          0.048288
mean symmetry                0.181552
mean fractal dimension       0.062889
radius error                 0.401108
texture error                1.223169
perimeter error              2.837574
area error                  39.672730
smoothness error             0.007096
compactness error            0.025565
concavity error              0.032191
concave points error         0.011840
symmetry error               0.020756
fractal dimension error      0.003832
worst radius                16.121150
worst texture               25.663828
worst perimeter            106.244395
worst area                 863.284766
worst smoothness             0.132500
worst compactness            0.253774
worst concav

In [344]:
#Splitting the Input into Training(90%) and Testing(10%) sets and using Stratify to push same number of labels into  the training and testing sets and using random_state to fix particular seed value so that the same record is pushed into the Training and Testing Sets at each execution 
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.1, stratify = Y, random_state = 3)
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (512,) (57,)


Splitting the Input into Training(80%) and Testing(20%) sets

In [345]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,stratify = Y, random_state = 2)
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (455,) (114,)


In [346]:
#Binarization of Input Feature Vector for both training and testing sets using pd.cut function 
X_binarised_train = X_train.apply(pd.cut, bins = 2, labels=[1,0])
X_binarised_test = X_test.apply(pd.cut, bins=2, labels=[1,0])


In [347]:
#Binarised Input Feature Vector of Training set
print(X_binarised_train)

    mean radius mean texture mean perimeter mean area mean smoothness  \
23            0            1              0         0               1   
316           1            1              1         1               1   
251           1            1              1         1               1   
391           1            1              1         1               1   
444           0            1              0         1               1   
..          ...          ...            ...       ...             ...   
71            1            1              1         1               1   
233           0            0              0         1               1   
346           1            1              1         1               1   
17            1            1              1         1               0   
75            1            1              1         1               1   

    mean compactness mean concavity mean concave points mean symmetry  \
23                 1              1               

In [348]:
#Binarised Input Feature Vector of Testing set
print(X_binarised_test)
type(X_binarised_train)

    mean radius mean texture mean perimeter mean area mean smoothness  \
223           1            1              1         1               1   
418           1            1              1         1               1   
246           1            1              1         1               1   
387           1            1              1         1               1   
407           1            1              1         1               1   
..          ...          ...            ...       ...             ...   
467           1            1              1         1               1   
108           0            1              0         0               0   
430           1            0              1         1               1   
129           0            0              0         0               1   
188           1            1              1         1               1   

    mean compactness mean concavity mean concave points mean symmetry  \
223                1              1               

pandas.core.frame.DataFrame

In [349]:
#Converting the Train and Test Feature Vector datasets from DataFrame type to numpy array 
X_binarised_train = X_binarised_train.values
X_binarised_test = X_binarised_test.values

In [350]:
type(X_binarised_train)


numpy.ndarray

In [351]:
# What are the possible values that b(threshold) can attain for the given dataset 
#checking for b(threshold) value as 10
b=10
#importing randit from random to generate random values from within the range of the values of the binarised Feature Vector of the training set
from random import randint
i = randint(0,X_binarised_train.shape[0])

# Building the logic for our ML classification Model with the training data as input feeding into it 
if(np.sum(X_binarised_train[i,:])>=b):
    print('Model Prediction is Malignant')
else:
    print('Model prediction is Benign')
    
#Building the logic to check whether Model Prediction of Label is same and accurate to the Target Variable Value for particular Feature Vector(X)    
if(Y_train[i]==1):
    print('Actual Outcome is Malignant')
else:
    print('Actual Outcome is Benign')

Model Prediction is Malignant
Actual Outcome is Benign


In [352]:
# Accuracy = (No. of rows predicted correctly / total no. of rows)
#Accuray Calculation with all possible values of b (ranging from 1 to 30) for the Training Set

for b in range(X_binarised_train.shape[1]+1):
    accurate_rows = 0
    
    for x,y in zip(X_binarised_train, Y_train):
        y_pred = (np.sum(x)>=b)
        accurate_rows+= (y == y_pred)
  
    print('For b= ',b ,'on training set, Accurate rows: ', accurate_rows , 'Accuracy= ',accurate_rows/X_binarised_train.shape[0])



For b=  0 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  1 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  2 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  3 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  4 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  5 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  6 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  7 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  8 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  9 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  10 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  11 on training set, Accurate rows:  285 Accuracy=  0.6263736263736264
For b=  12 on training set, Accurate rows:  285 Accuracy=  0.6263736263736

Observation:- It can be noticed that for the value of b(Threshold) = 27, the ML Classification Model predicts Labels with the Highest Accuracy of 0.8483516483516483

In [353]:
#Accuray Calculation with the derived optimum value of b = 27 for the Testing  Set
acc_rows=0
t=X_binarised_test.shape[0]
b=27
for x,y in zip(X_binarised_test, Y_test):
        y_pred = (np.sum(x)>=b)
        acc_rows+= (y == y_pred)
    
print('For b= ',b ,'on testing set, Accurate rows: ', acc_rows , 'Accuracy= ',acc_rows/t)



For b=  27 on testing set, Accurate rows:  103 Accuracy=  0.9035087719298246


Testing Vs Traing Accuracy for optimum value of b=27
For b=  27 on training set, Accurate rows:  386 Accuracy=  0.8483516483516483
For b=  27 on testing set, Accurate rows:  103 Accuracy=  0.9035087719298246

Splitting the Input into Training(70%) and Testing(30%) sets 

In [354]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.3, stratify = Y, random_state = 3)
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (398,) (171,)


In [355]:
#Binarization of Input Feature Vector for both training and testing sets using pd.cut function 
X_binarised_train = X_train.apply(pd.cut, bins = 2, labels=[1,0])
X_binarised_test = X_test.apply(pd.cut, bins=2, labels=[1,0])


In [356]:
#Binarised Input Feature Vector of Training set
print(X_binarised_train)

    mean radius mean texture mean perimeter mean area mean smoothness  \
387           1            1              1         1               1   
225           1            1              1         1               1   
430           1            1              1         1               1   
324           1            1              1         1               1   
314           1            1              1         1               1   
..          ...          ...            ...       ...             ...   
166           1            1              1         1               1   
463           1            1              1         1               1   
358           1            1              1         1               1   
381           1            1              1         1               1   
127           0            1              0         1               1   

    mean compactness mean concavity mean concave points mean symmetry  \
387                1              1               

In [357]:
#Binarised Input Feature Vector of Testing set
print(X_binarised_test)
type(X_binarised_train)

    mean radius mean texture mean perimeter mean area mean smoothness  \
153           1            1              1         1               1   
48            1            1              1         1               0   
163           1            0              1         1               1   
371           1            1              1         1               1   
340           1            1              1         1               1   
..          ...          ...            ...       ...             ...   
106           1            1              1         1               0   
280           0            0              0         1               1   
491           0            1              1         1               1   
24            1            1              1         1               0   
515           1            1              1         1               0   

    mean compactness mean concavity mean concave points mean symmetry  \
153                1              1               

pandas.core.frame.DataFrame

In [358]:
#Converting the Train and Test Feature Vector datasets from DataFrame type to numpy array 
X_binarised_train = X_binarised_train.values
X_binarised_test = X_binarised_test.values

In [359]:
type(X_binarised_train)

numpy.ndarray

In [360]:
# Accuracy = (No. of rows predicted correctly / total no. of rows)
#Accuray Calculation with all possible values of b (ranging from 1 to 30) for the Training Set 
for b in range(X_binarised_train.shape[1]+1):
    accurate_rows = 0
    
    for x,y in zip(X_binarised_train, Y_train):
        y_pred = (np.sum(x)>=b)
        accurate_rows+= (y == y_pred)
  
    print('For b= ',b ,'on training set, Accurate rows: ', accurate_rows , 'Accuracy= ',accurate_rows/X_binarised_train.shape[0])



For b=  0 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  1 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  2 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  3 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  4 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  5 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  6 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  7 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  8 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  9 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  10 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  11 on training set, Accurate rows:  250 Accuracy=  0.628140703517588
For b=  12 on training set, Accurate rows:  252 Accuracy=  0.6331658291457286
For b=  

Observation:- It can be noticed that for the value of b(Threshold) = 27, the ML Classification Model predicts Labels with the Highest Accuracy of 0.8517587939698492 

In [361]:
#Accuray Calculation with the derived optimum value of b = 27 for the Testing  Set
acc_rows=0
t=X_binarised_test.shape[0]
b=27
for x,y in zip(X_binarised_test, Y_test):
        y_pred = (np.sum(x)>=b)
        acc_rows+= (y == y_pred)
    
print('For b= ',b ,'on testing set, Accurate rows: ', acc_rows , 'Accuracy= ',acc_rows/t)


For b=  27 on testing set, Accurate rows:  148 Accuracy=  0.8654970760233918


Testing Vs Traing Accuracy for optimum value of  b=27
For b=  27 on training set, Accurate rows:  339 Accuracy=  0.8517587939698492
For b=  27 on testing set, Accurate rows:  148 Accuracy=  0.8654970760233918